Property right of AssignmentExpression expected node to be of a type ["Expression"] but instead got null

<!---
Thanks for filing an issue 😄 ! Before you submit, please read the following:

Search open/closed issues before submitting since someone might have asked the same thing before!
-->

# 🐛 bug report

I wanted to try out new tree shaking feature at 1.9.0 but the project don't build when I add that cli command.

## 🎛 Configuration (.babelrc, package.json, cli command)

`{
  "env": {
    "production": {
      "presets": ["env", "react", "flow"],
      "plugins": [
        "emotion",
        [
          "transform-runtime",
          {
            "polyfill": false,
            "regenerator": true
          }
        ],
        "transform-decorators-legacy",
        "transform-class-properties",
        ["import", { "libraryName": "antd", "libraryDirectory": "es", "style": "css"}]
      ]
    },
    "development": {
      "presets": ["env", "react", "flow"],
      "plugins": [
        "emotion",
        [
          "transform-runtime",
          {
            "polyfill": false,
            "regenerator": true
          }
        ],
        "transform-decorators-legacy",
        "transform-class-properties",
        ["import", { "libraryName": "antd", "libraryDirectory": "es", "style": "css"}]
      ]
    },
    "test": {
      "presets": ["env", "react", "flow"],
      "plugins": [
        "emotion",
        [
          "transform-runtime",
          {
            "polyfill": false,
            "regenerator": true
          }
        ],
        "transform-decorators-legacy",
        "transform-class-properties"
      ]
    }
  }
}`

## 🤔 Expected Behavior

It should build as when I use parcel build without this tree shaking command

## 😯 Current Behavior

<!--- Tell us what happens instead of the expected behavior -->

It does not finish the build.

<!--- If you are seeing an error, please include the full error message and stack trace -->

This is the error message:

Property right of AssignmentExpression expected node to be of a type ["Expression"] but instead got null

I also included the error log file

## 💁 Possible Solution

<!--- Not obligatory, but suggest a fix/reason for the bug -->

## 🔦 Context

<!--- How has this issue affected you? What are you trying to accomplish? -->

I want to try out new features in Parcel.

<!--- Providing context helps us come up with a solution that is most useful in the real world -->

## 💻 Code Sample

<!-- Please provide a code repository, gist, code snippet or sample files to reproduce the issue -->

## 🌍 Your Environment

<!--- Include as many relevant details about the environment you experienced the bug in -->

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | 1.9.0
| Node             | 8.11.1
| npm/Yarn         | 6.0.0
| Operating System | Windows 10 x64

<!-- Love parcel? Please consider supporting our collective:
👉  https://opencollective.com/parcel/donate -->
[yarn-error.log](https://github.com/parcel-bundler/parcel/files/2105611/yarn-error.log)

It gives the log below at the terminal. I don't understand which file it is from looking at this log. Can you help me spotting the file please


`C:\Users\HP\Desktop\HardwareWeb>yarn build
yarn run v1.5.1
$ npm run rm:build && cross-env NODE_ENV=production concurrently "npm run babel:build" "npm run parcel:build"

> azuretest@1.0.0 rm:build C:\Users\HP\Desktop\HardwareWeb
> rimraf build

[1]
[1] > azuretest@1.0.0 parcel:build C:\Users\HP\Desktop\HardwareWeb
[1] > parcel build src/index.html -d build/public --experimental-scope-hoisting
[1]
[0]
[0] > azuretest@1.0.0 babel:build C:\Users\HP\Desktop\HardwareWeb
[0] > babel src/server -d build
[0]
∞  Building index.html...
src\server\api_router.js -> build\api_router.js
[0] src\server\database\database.js -> build\database\database.js
[0] src\server\database\db_content.js -> build\database\db_content.js
[0] src\server\database\db_playlist.js -> build\database\db_playlist.js
[0] src\server\database\db_screen.js -> build\database\db_screen.js
∞  Building index.js...
src\server\database\models.js -> build\database\models.js
[0] src\server\express.js -> build\express.js
×  Property right of AssignmentExpression expected node to be of a type ["Expression"] but instead got null
    at Object.validate (C:\Users\HP\Desktop\HardwareWeb\node_modules\babel-types\lib\definitions\index.js:109:13)
[1]     at Object.validate (C:\Users\HP\Desktop\HardwareWeb\node_modules\babel-types\lib\index.js:505:9)
[1]     at NodePath._replaceWith (C:\Users\HP\Desktop\HardwareWeb\node_modules\babel-traverse\lib\path\replacement.js:176:7)
[1]     at NodePath._remove (C:\Users\HP\Desktop\HardwareWeb\node_modules\babel-traverse\lib\path\removal.js:58:10)
[1]     at NodePath.remove (C:\Users\HP\Desktop\HardwareWeb\node_modules\babel-traverse\lib\path\removal.js:30:8)
[1]     at binding.referencePaths.concat.forEach.path (C:\Users\HP\Desktop\HardwareWeb\node_modules\parcel-bundler\src\scope-hoisting\shake.js:41:22)
[1]     at Array.forEach (<anonymous>)
[1]     at Object.keys.forEach.name (C:\Users\HP\Desktop\HardwareWeb\node_modules\parcel-bundler\src\scope-hoisting\shake.js:32:10)
[1]     at Array.forEach (<anonymous>)
[1]     at treeShake (C:\Users\HP\Desktop\HardwareWeb\node_modules\parcel-bundler\src\scope-hoisting\shake.js:20:33)
[1] npm ERR! code ELIFECYCLE
[1] npm ERR! errno 1
[1] npm ERR! azuretest@1.0.0 parcel:build: `parcel build src/index.html -d build/public --experimental-scope-hoisting`
[1] npm ERR! Exit status 1
[1] npm ERR!
npm ERR! Failed at the azuretest@1.0.0 parcel:build script.
[1] npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
[1]
[1] npm ERR! A complete log of this run can be found in:
[1] npm ERR!     C:\Users\HP\AppData\Roaming\npm-cache\_logs\2018-06-15T09_49_40_650Z-debug.log
[1] npm run parcel:build exited with code 1
error An unexpected error occurred: "Command failed.
Exit code: 1
Command: C:\\WINDOWS\\system32\\cmd.exe
Arguments: /d /s /c npm run rm:build && cross-env NODE_ENV=production concurrently \"npm run babel:build\" \"npm run parcel:build\"
Directory: C:\\Users\\HP\\Desktop\\HardwareWeb
Output:
".
info If you think this is a bug, please open a bug report with the information provided in "C:\\Users\\HP\\Desktop\\HardwareWeb\\yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
`

--- a/src/scope-hoisting/shake.js
+++ b/src/scope-hoisting/shake.js
@@ -29,28 +29,7 @@ function treeShake(scope) {
       binding.path.remove();
       binding.referencePaths
         .concat(binding.constantViolations)
-        .forEach(path => {
-          if (path.parentPath.isMemberExpression()) {
-            let parent = path.parentPath.parentPath;
-            if (
-              parent.parentPath.isSequenceExpression() &&
-              parent.parent.expressions.length === 1
-            ) {
-              parent.parentPath.remove();
-            } else if (!parent.removed) {
-              parent.remove();
-            }
-          } else if (isUnusedWildcard(path) && !path.parentPath.removed) {
-            path.parentPath.remove();
-          } else if (path.isAssignmentExpression()) {
-            let parent = path.parentPath;
-            if (!parent.isExpressionStatement()) {
-              path.replaceWith(path.node.right);
-            } else {
-              path.remove();
-            }
-          }
-        });
+        .forEach(safeRemove);
 
       scope.removeBinding(name);
       removed = true;
@@ -59,6 +38,53 @@ function treeShake(scope) {
 }
 
 module.exports = treeShake;
+
+function getExpressionValue(path) {
+  if(path.isAssignmentExpression()) {
+    return getExpressionValue(path.get('right'))
+  }
+
+  return path
+}
+
+function safeRemove(path) {
+  if(path.removed) {
+    return
+  }
+
+  let {parentPath} = path
+
+  if(isUnusedWildcard(path)) {
+    return parentPath.remove()
+  }
+
+  if(parentPath.isMemberExpression()) {
+    return safeRemove(parentPath)
+  }
+
+  if(parentPath.isAssignmentExpression()) {
+    let other = parentPath.get('right')
+
+    if(other === parentPath) {
+      other = parentPath.get('left')
+    }
+
+    return parentPath.replaceWith(other)
+  }
+
+  if(path.isExpression()) {
+    let value = getExpressionValue(path)
+
+    if(value === path) {
+      value = t.nullLiteral()
+    }
+
+    path.replaceWith(value)
+  }
+  else {
+    path.remove()
+  }
+}
 
 // Check if a binding is safe to remove and returns it if it is.
 function getUnusedBinding(path, name) {

--- a/src/scope-hoisting/shake.js
+++ b/src/scope-hoisting/shake.js
@@ -27,9 +27,7 @@ function treeShake(scope) {
 
       // Remove the binding and all references to it.
       binding.path.remove();
-      binding.referencePaths
-        .concat(binding.constantViolations)
-        .forEach(safeRemove);
+      binding.referencePaths.concat(binding.constantViolations).forEach(remove);
 
       scope.removeBinding(name);
       removed = true;
@@ -38,53 +36,6 @@ function treeShake(scope) {
 }
 
 module.exports = treeShake;
-
-function getExpressionValue(path) {
-  if(path.isAssignmentExpression()) {
-    return getExpressionValue(path.get('right'))
-  }
-
-  return path
-}
-
-function safeRemove(path) {
-  if(path.removed) {
-    return
-  }
-
-  let {parentPath} = path
-
-  if(isUnusedWildcard(path)) {
-    return parentPath.remove()
-  }
-
-  if(parentPath.isMemberExpression()) {
-    return safeRemove(parentPath)
-  }
-
-  if(parentPath.isAssignmentExpression()) {
-    let other = parentPath.get('right')
-
-    if(other === parentPath) {
-      other = parentPath.get('left')
-    }
-
-    return parentPath.replaceWith(other)
-  }
-
-  if(path.isExpression()) {
-    let value = getExpressionValue(path)
-
-    if(value === path) {
-      value = t.nullLiteral()
-    }
-
-    path.replaceWith(value)
-  }
-  else {
-    path.remove()
-  }
-}
 
 // Check if a binding is safe to remove and returns it if it is.
 function getUnusedBinding(path, name) {
@@ -149,4 +100,25 @@ function isUnusedWildcard(path) {
     // check if the $id$exports variable is used
     !getUnusedBinding(path, parent.arguments[1].name)
   );
+}
+
+function remove(path) {
+  if (path.isAssignmentExpression()) {
+    if (!path.parentPath.isExpressionStatement()) {
+      path.replaceWith(path.node.right);
+    } else {
+      path.remove();
+    }
+  } else if (isExportAssignment(path)) {
+    remove(path.parentPath.parentPath);
+  } else if (isUnusedWildcard(path)) {
+    remove(path.parentPath);
+  } else if (
+    path.parentPath.isSequenceExpression() &&
+    path.parent.expressions.length === 1
+  ) {
+    remove(path.parentPath);
+  } else if (!path.removed) {
+    path.remove();
+  }
 }

--- a/test/integration/scope-hoisting/commonjs/export-local/a.js
+++ b/test/integration/scope-hoisting/commonjs/export-local/a.js
@@ -1,0 +1,1 @@
+module.exports = require('./b').foo;

--- a/test/integration/scope-hoisting/commonjs/export-local/b.js
+++ b/test/integration/scope-hoisting/commonjs/export-local/b.js
@@ -1,0 +1,1 @@
+var x = exports.foo = exports.bar = 3;

--- a/test/scope-hoisting.js
+++ b/test/scope-hoisting.js
@@ -766,5 +766,17 @@ describe('scope hoisting', function() {
       assert(contents.includes('foo'));
       assert(!contents.includes('bar'));
     });
+
+    it('supports removing an unused inline export with uglify minification', async function() {
+      // Uglify does strange things to multiple assignments in a line.
+      // See https://github.com/parcel-bundler/parcel/issues/1549
+      let b = await bundle(
+        __dirname + '/integration/scope-hoisting/commonjs/export-local/a.js',
+        {minify: true}
+      );
+
+      let output = await run(b);
+      assert.deepEqual(output, 3);
+    });
   });
 });

